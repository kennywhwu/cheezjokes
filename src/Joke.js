import React, { Component } from 'react';
import './Joke.css';

class Joke extends Component {
  // static defaultProps = {
  //   votes: 0
  // };

  render() {
    return (
      <div className="Joke">
        <span>{this.props.joke}</span>
        <i
          className="up fas fa-thumbs-up"
          onClick={() => this.props.handleClick('up', this.props.id)}
        />{' '}
        <i
          className="down fas fa-thumbs-down"
          onClick={() => this.props.handleClick('down', this.props.id)}
        />{' '}
        <i className="vote-total">{this.props.votes}</i>
      </div>
    );
  }
}

export default Joke;
