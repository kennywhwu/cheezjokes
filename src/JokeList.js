import React, { Component } from 'react';
import './JokeList.css';
import Joke from './Joke';
import axios from 'axios';
// import uuid from 'uuid/v4';

class JokeList extends Component {
  constructor(props) {
    super(props);
    this.state = { jokes: [], isLoading: true };
    this.vote = this.vote.bind(this);
    this.refreshList = this.refreshList.bind(this);
  }

  async componentDidMount() {
    if (JSON.parse(localStorage.getItem('jokes'))) {
      this.setState({
        jokes: JSON.parse(localStorage.getItem('jokes')),
        isLoading: false
      });
    } else {
      let jokes = [];
      for (let i = 0; i < 10; i++) {
        await this.requestJoke(jokes);
      }
      this.setState({ jokes, isLoading: false });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    localStorage.setItem('jokes', JSON.stringify(this.state.jokes));
  }

  async requestJoke(jokes) {
    let response = await axios.get('https://icanhazdadjoke.com/', {
      headers: { Accept: 'application/json' }
    });
    let joke = { joke: response.data.joke, id: response.data.id, votes: 0 };

    if (jokes.every((e, i) => e.id !== joke.id)) {
      jokes.push(joke);
    } else {
      await this.requestJoke(jokes);
    }
  }

  vote(direction, id) {
    console.log(direction, id);
    let newJoke = {
      joke: this.state.jokes[id].joke,
      votes: this.state.jokes[id].votes,
      id: this.state.jokes[id].id
    };

    // Change vote total based on direction
    if (direction === 'up') {
      newJoke.votes += 1;
    } else {
      newJoke.votes -= 1;
    }

    // Pull in jokes array from current state
    let votedJokes = {
      jokes: [
        ...this.state.jokes.slice(0, id),
        newJoke,
        ...this.state.jokes.slice(id + 1)
      ]
    };

    // Order jokes based on reverse order vote
    let orderedJokes = votedJokes.jokes.sort((a, b) => b.votes - a.votes);

    // Set state
    this.setState(st => ({
      jokes: orderedJokes
    }));
  }

  async refreshList() {
    localStorage.removeItem('jokes');
    this.setState({ isLoading: true });
    let jokes = [];
    for (let i = 0; i < 10; i++) {
      await this.requestJoke(jokes);
    }
    this.setState({ jokes, isLoading: false });
  }

  render() {
    let result;
    let jokes;
    if (!this.state.isLoading) {
      jokes = this.state.jokes.map((e, i) => (
        <Joke
          joke={e.joke}
          key={e.id}
          id={i}
          handleClick={this.vote}
          votes={e.votes}
        />
      ));
      result = (
        <div>
          {jokes}
          <button onClick={this.refreshList}>Refresh Jokes</button>
        </div>
      );
    } else {
      result = (
        <h1>It's loading already, can you just wait a minute please???</h1>
      );
    }

    return <div className="JokeList">{result}</div>;
  }
}

export default JokeList;
