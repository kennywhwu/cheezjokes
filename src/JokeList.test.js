/* React tests for JokeList component */

import React from 'react';
import { shallow, mount } from 'enzyme';
import JokeList from './JokeList';
import Joke from './Joke';
import toJson from 'enzyme-to-json';

// Smoke tests
it('should render without children without crashing', function() {
  shallow(<JokeList />);
});

it('should render with children without crashing', function() {
  mount(<JokeList />);
});

// Snapshot test
it('should match snapshot', function() {
  let wrapper = shallow(<JokeList />);
  // wrapper.setState({ vote: });
  let serialized = toJson(wrapper);
  expect(serialized).toMatchSnapshot();
});

// // Simulation test
it('should display completed story when all prompts are submitted', function() {
  let wrapper = shallow(<JokeList />);
  wrapper.setState({
    jokes: [{ joke: 'hi', votes: 0 }],
    isLoading: false
  });
  let elem = wrapper.find('Joke').first();
  expect(elem.matchesElement(<Joke joke="hi" />)).toBe(true);
});

it('should display the correct number of votes based on up-vote/down-vote clicks', function() {
  let wrapper = mount(<JokeList />);
  wrapper.setState({
    jokes: [{ joke: 'hi', votes: 0 }],
    isLoading: false
  });
  wrapper
    .find('i.up')
    .first()
    .simulate('click');
  let elem = wrapper.find('i.vote-total').first();
  expect(elem.text()).toBe('1');
  wrapper
    .find('i.up')
    .first()
    .simulate('click');
  elem = wrapper.find('i.vote-total').first();
  expect(elem.text()).toBe('2');
  wrapper
    .find('i.down')
    .first()
    .simulate('click');
  elem = wrapper.find('i.vote-total').first();
  expect(elem.text()).toBe('1');
});

it('should display jokes in the right order based on vote total', function() {
  let wrapper = mount(<JokeList />);
  wrapper.setState({
    jokes: [{ joke: 'hi', votes: 1 }, { joke: 'bye', votes: 1 }],
    isLoading: false
  });
  wrapper
    .find('i.down')
    .first()
    .simulate('click');
  let elem = wrapper.find('span').first();
  expect(elem.text()).toBe('bye');
});
