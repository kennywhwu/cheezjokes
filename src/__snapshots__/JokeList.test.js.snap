// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="JokeList"
>
  <h1>
    It's loading already, can you just wait a minute please???
  </h1>
</div>
`;
